packages/core/src/asyncContext/stackStrategy.ts

```diff
@@ -0,0 +1,178 @@
+
+  public constructor(scope?: ScopeInterface, isolationScope?: ScopeInterface) {
+    let assignedScope;
+    if (!scope) {
+      assignedScope = new Scope();
+    } else {
+      assignedScope = scope;
+    }
+
+    let assignedIsolationScope;
+    if (!isolationScope) {
+      assignedIsolationScope = new Scope();
+    } else {
+      assignedIsolationScope = isolationScope;
+    }
+
+    this._stack = [{ scope: assignedScope }];
+    this._isolationScope = assignedIsolationScope;
+  }
+
+  /**
+   * Fork a scope for the stack.
+   */
+  public withScope<T>(callback: (scope: ScopeInterface) => T): T {
+    const scope = this._pushScope();
+
+    let maybePromiseResult: T;
+    try {
+      maybePromiseResult = callback(scope);
+    } catch (e) {
+      this._popScope();
+      throw e;
+    }
+
+    if (isThenable(maybePromiseResult)) {
+      // @ts-expect-error - isThenable returns the wrong type
+      return maybePromiseResult.then(
+        res => {
+          this._popScope();
+          return res;
+        },
+        e => {
+          this._popScope();
+          throw e;
+        },
+      );
+    }
+
+    this._popScope();
+    return maybePromiseResult;
+  }
+
+  /**
+   * Get the client of the stack.
+   */
+  public getClient<C extends Client>(): C | undefined {
+    return this.getStackTop().client as C;
+  }
+
+  /**
+   * Returns the scope of the top stack.
+   */
+  public getScope(): ScopeInterface {
+    return this.getStackTop().scope;
+  }
+
+  /**
+   * Get the isolation scope for the stack.
+   */
+  public getIsolationScope(): ScopeInterface {
+    return this._isolationScope;
+  }
+
+  /**
+   * Returns the scope stack for domains or the process.
+   */
+  public getStack(): Layer[] {
+    return this._stack;
+  }
+
+  /**
+   * Returns the topmost scope layer in the order domain > local > process.
+   */
+  public getStackTop(): Layer {
+    return this._stack[this._stack.length - 1];
+  }
+
+  /**
+   * Push a scope to the stack.
+   */
+  private _pushScope(): ScopeInterface {
+    // We want to clone the content of prev scope
+    const scope = this.getScope().clone();
+    this.getStack().push({
+      client: this.getClient(),
+      scope,
+    });
+    return scope;
+  }
+
+  /**
+   * Pop a scope from the stack.
+   */
+  private _popScope(): boolean {
+    if (this.getStack().length <= 1) return false;
+    return !!this.getStack().pop();
+  }
+}
+
+/**
+ * Get the global async context stack.
+ * This will be removed during the v8 cycle and is only here to make migration easier.
+ */
+function getAsyncContextStack(): AsyncContextStack {
+  const registry = getMainCarrier();
+  const sentry = getSentryCarrier(registry) as { hub?: AsyncContextStack };
+
+  // If there's no hub, or its an old API, assign a new one
+  if (sentry.hub) {
+    return sentry.hub;
+  }
+
+  sentry.hub = new AsyncContextStack(getDefaultCurrentScope(), getDefaultIsolationScope());
+  return sentry.hub;
```

Comment:
can we stop referencing it by `.hub` here?

Reply:
sadly not because some stuff like the loader depends on this 😬 @Lms24 learned this the hard way I believe?

Reply:
that's also the reason that this still has a `getClient()` method!

Reply:
yes, if we remove/rename `.hub` our loader tests will fail because our loader script accesses `__SENTRY__.hub.getClient()` to check if the SDK was initialized before the loader otherwise initializes the SDK (at least that's what I think it does; the script certainly makes the access). So I guess we can only remove this once we update the loader script 🤔

Reply:
we can still do this when v8 becomes stable I guess, but it's possibly not worth it too as it may break stuff weirdly I guess that depends on internals 😅

Reply:
could we leave a comment about this somewhere?

Reply:
See https://github.com/getsentry/sentry-javascript/pull/11633 👍